daphne/src/messages.rs

```diff
@@ -809,6 +822,266 @@ impl Decode for HpkeCiphertext {
+            DpMechanism::NotImplemented(x) => x,
+        }
+    }
+}
+
+impl Encode for DpMechanism {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        u8::from(*self).encode(bytes);
+    }
+}
+
+impl Decode for DpMechanism {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        match u8::decode(bytes)? {
+            x if x == DP_MECHANISM_RESERVED => Ok(Self::Reserved),
+            x if x == DP_MECHANISM_NONE => Ok(Self::None),
+            x => Ok(Self::NotImplemented(x)),
+        }
+    }
+}
+
+// The Differential Privacy configuration defined in draft-wang-ppm-dap-taskprov-00
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct DpConfig {
+    pub mechanism: DpMechanism,
+}
+
+impl Encode for DpConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.mechanism.encode(bytes);
+    }
+}
+
+impl Decode for DpConfig {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            mechanism: DpMechanism::decode(bytes)?,
+        })
+    }
+}
+
+// The VDAF configuration defined in draft-wang-ppm-dap-taskprov-00
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct VdafConfig {
+    pub dp_config: DpConfig,
+    pub vdaf_type: VdafTypeVar,
+}
+
+impl Encode for VdafConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.dp_config.encode(bytes);
+        self.vdaf_type.encode(bytes);
+    }
+}
+
+impl Decode for VdafConfig {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            dp_config: DpConfig::decode(bytes)?,
+            vdaf_type: VdafTypeVar::decode(bytes)?,
+        })
+    }
+}
+
+// URL encode / decode helper struct
+pub(crate) struct UrlBytes {
+    bytes: Vec<u8>,
+}
+
+impl Encode for UrlBytes {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        encode_u16_items(bytes, &(), &self.bytes)
+    }
+}
+
+impl Decode for UrlBytes {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            bytes: decode_u16_items(&(), bytes)?,
+        })
+    }
+}
+
+// The variable-length part of a QueryConfig in draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub enum QueryConfigVar {
+    TimeInterval,
+    FixedSize { max_batch_size: u32 },
+}
+
+// There is no Encode or Decode for QueryConfigVar as we have to split the query type and
+// the variable length stuff in the message format, so we do all of the work in QueryConfig's
+// Encode and Decode.
+
+// The query configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct QueryConfig {
+    pub time_precision: Duration,
+    pub max_batch_query_count: u16,
+    pub min_batch_size: u32,
+    pub var: QueryConfigVar,
+}
+
+impl Encode for QueryConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        match &self.var {
```

Comment:
Yeah it's super awkward that we have to match on `var` twice. Consider filing an issue against the spec to fix this.

Reply:
This will be addressed in taskprov-01 by https://github.com/wangshan/draft-wang-ppm-dap-taskprov/pull/28.